File: lib/Models/dataset.dart
// ignore_for_file: public_member_api_docs, sort_constructors_first
class Dataset {
  String refNo;
  String bankName;
  String branchName;
  String partyName;
  String colonyName;
  String cityVillageName;
  double latitude;
  double longitude;
  int marketRate;
  String unit;
  DateTime createdAt;
  String remarks;
  String colorMark;
  DateTime dateOfVisit;
  int id;
  String entryBy;

  Dataset({
    required this.refNo,
    required this.bankName,
    required this.branchName,
    required this.partyName,
    required this.colonyName,
    required this.cityVillageName,
    required this.latitude,
    required this.longitude,
    required this.marketRate,
    required this.unit,
    required this.createdAt,
    required this.remarks,
    required this.colorMark,
    required this.dateOfVisit,
    required this.id,
    required this.entryBy,
  });

  Dataset copyWith(
    Map<String, dynamic> updates, {
    String? refNo,
    String? bankName,
    String? branchName,
    String? partyName,
    String? colonyName,
    String? cityVillageName,
    double? latitude,
    double? longitude,
    int? marketRate,
    String? unit,
    DateTime? createdAt,
    String? remarks,
    int? id,
    DateTime? dateOfVisit,
    String? colorMark,
    String? entryBy,
  }) {
    return Dataset(
      refNo: refNo ?? this.refNo,
      bankName: bankName ?? this.bankName,
      branchName: branchName ?? this.branchName,
      partyName: partyName ?? this.partyName,
      colonyName: colonyName ?? this.colonyName,
      cityVillageName: cityVillageName ?? this.cityVillageName,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      marketRate: marketRate ?? this.marketRate,
      unit: unit ?? this.unit,
      createdAt: createdAt ?? this.createdAt,
      remarks: remarks ?? this.remarks,
      id: id ?? this.id,
      dateOfVisit: dateOfVisit ?? this.dateOfVisit,
      colorMark: colorMark ?? this.colorMark,
      entryBy: entryBy ?? this.entryBy,
    );
  }

  factory Dataset.fromJson(Map<String, dynamic> map) {
    return Dataset(
      refNo: map['refNo'] ?? '',
      bankName: map['bankName'] ?? '',
      branchName: map['branchName'] ?? '',
      partyName: map['partyName'] ?? '',
      colonyName: map['colonyName'] ?? '',
      cityVillageName: map['cityVillageName'] ?? '',
      latitude: map['latitude'] ?? 0.0,
      longitude: map['longitude'] ?? 0.0,
      marketRate: map['marketRate'] ?? 0,
      unit: map['unit'] ?? '',
      createdAt: DateTime.tryParse(map['createdAt'] ?? '') ?? DateTime.now(),
      remarks: map['remarks'] ?? '',
      id: map['id'],
      dateOfVisit:
          DateTime.tryParse(map['dateOfVisit'] ?? '') ?? DateTime.now(),
      colorMark: map['colorMark'] ?? '',
      entryBy: map['entryBy'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'refNo': refNo,
      'bankName': bankName,
      'branchName': branchName,
      'partyName': partyName,
      'colonyName': colonyName,
      'cityVillageName': cityVillageName,
      'latitude': latitude,
      'longitude': longitude,
      'marketRate': marketRate,
      'createdAt': createdAt.toIso8601String(), // Convert to ISO string
      'remarks': remarks,
      'id': id,
      'dateOfVisit': dateOfVisit.toIso8601String(), // Convert to ISO string
      'colorMark': colorMark,
      'entryBy': entryBy,
    };
  }
}

File: server/routes/dataset.js
const express = require("express");
const auth = require("../middlewares/auth");
const { PrismaClient } = require("@prisma/client");

const datasetRouter = express.Router();
const prisma = new PrismaClient();

datasetRouter.get("/api/datasets", async (req, res) => {
  try {
    const datasets = await prisma.dataset.findMany();

    res.status(201).json(datasets);
  } catch (e) {
    res.status(500).json({ error: e.message });
  }
});

datasetRouter.post("/api/create-dataset", async (req, res) => {
  try {
    const {
      refNo,
      bankName,
      branchName,
      partyName,
      colonyName,
      cityVillageName,
      latitude,
      longitude,
      marketRate,
      unit,
      createdAt,
      remarks,
      colorMark,
      dateOfVisit,
      entryBy,
    } = req.body;

    const existingDataset = await prisma.dataset.findFirst({
      where: { latitude: latitude },
    });

    if (existingDataset) {
      return res
        .status(400)
        .json({ msg: "Dataset with the same latitude already exists!" });
    }

    const parsedCreatedAt = new Date(createdAt);
    const parsedDateOfVisit = new Date(dateOfVisit);

    const createdDataset = await prisma.dataset.create({
      data: {
        refNo: refNo,
        bankName: bankName,
        branchName: branchName,
        partyName: partyName,
        colonyName: colonyName,
        cityVillageName: cityVillageName,
        latitude: latitude,
        longitude: longitude,
        marketRate: marketRate,
        unit: unit,
        createdAt: parsedCreatedAt,
        remarks: remarks,
        colorMark: colorMark,
        dateOfVisit: parsedDateOfVisit,
        entryBy: entryBy,
      },
    });

    res.status(201).json(createdDataset);
  } catch (e) {
    res.status(500).json({ error: e.message });
  }
});

datasetRouter.delete("/api/delete-dataset/:id", async (req, res) => {
  try {
    const { id } = req.params;

    const dataset = await prisma.dataset.delete({
      where: {
        id: parseInt(id),
      },
    });

    res.status(201).json(dataset);
  } catch (e) {
    res.status(500).json({ error: e.message });
  }
});

datasetRouter.put("/api/update-dataset/:id", async (req, res) => {
  try {
    const { id } = req.params;
    const {
      refNo,
      bankName,
      branchName,
      partyName,
      colonyName,
      cityVillageName,
      latitude,
      longitude,
      marketRate,
      unit,
      remarks,
      colorMark,
    } = req.body;

    const updatedDataset = await prisma.dataset.update({
      where: {
        id: parseInt(id),
      },
      data: {
        refNo: refNo,
        bankName: bankName,
        branchName: branchName,
        partyName: partyName,
        colonyName: colonyName,
        cityVillageName: cityVillageName,
        latitude: latitude,
        longitude: longitude,
        marketRate: marketRate,
        unit: unit,
        remarks: remarks,
        colorMark: colorMark,
      },
    });

    res.status(201).json(updatedDataset);
  } catch (e) {
    res.status(500).json({ error: e.message });
  }
});

datasetRouter.put("/api/update-dataset-dateOfVisit/:id", async (req, res) => {
  try {
    const { id } = req.params;
    const { dateOfVisit } = req.body;

    const parsedDateOfVisit = new Date(dateOfVisit);

    const updatedDate = await prisma.dataset.update({
      where: {
        id: parseInt(id),
      },
      data: {
        dateOfVisit: parsedDateOfVisit,
      },
    });

    res.status(201).json(updatedDate);

  } catch (e) {
    res.status(500).json({ error: e.message });
  }
});

module.exports = datasetRouter;
